pkg/engine/runtime/terraform_runtime.go

```diff
@@ -0,0 +1,140 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/imdario/mergo"
+	"github.com/spf13/afero"
+	"kusionstack.io/kusion/pkg/engine/models"
+	"kusionstack.io/kusion/pkg/status"
+	"kusionstack.io/kusion/pkg/terraform"
+)
+
+var _ Runtime = &TerraformRuntime{}
+
+type TerraformRuntime struct {
+	terraform.WorkspaceStore
+}
+
+func NewTerraformRuntime() (Runtime, error) {
+	fs := afero.Afero{Fs: afero.NewOsFs()}
+	ws, err := terraform.GetWorkspaceStore(fs)
+	if err != nil {
+		return nil, err
+	}
+	TFRuntime := &TerraformRuntime{ws}
+	return TFRuntime, nil
+}
+
+// Apply terraform apply resource
+func (t *TerraformRuntime) Apply(ctx context.Context, request *ApplyRequest) *ApplyResponse {
+	planState := request.PlanResource
+	w, ok := t.Store[planState.ResourceKey()]
+	if !ok {
+		err := t.Create(ctx, planState)
+		if err != nil {
+			return &ApplyResponse{nil, status.NewErrorStatus(err)}
+		}
+		w = t.Store[planState.ResourceKey()]
+	}
+
+	// get terraform provider version
+	providerAddr, err := w.GetProvider()
+	if err != nil {
+		return &ApplyResponse{nil, status.NewErrorStatus(err)}
+	}
+
+	// terraform dry run merge state
+	if request.DryRun {
+		prior := request.PriorResource.DeepCopy()
+		if err := mergo.Merge(prior, planState, mergo.WithSliceDeepCopy, mergo.WithOverride); err != nil {
+			return &ApplyResponse{nil, status.NewErrorStatus(err)}
+		}
+
+		return &ApplyResponse{Resource: &models.Resource{
+			ID:         planState.ID,
+			Type:       planState.Type,
+			Attributes: prior.Attributes,
+			DependsOn:  planState.DependsOn,
+			Extensions: planState.Extensions,
+		}, Status: nil}
+	}
+	w.SetResource(planState)
+
+	if err := w.WriteHCL(); err != nil {
+		return &ApplyResponse{nil, status.NewErrorStatus(err)}
+	}
+
+	tfstate, err := w.Apply(ctx)
+	if err != nil {
+		return &ApplyResponse{nil, status.NewErrorStatus(err)}
+	}
+
+	r := terraform.ConvertTFState(tfstate, providerAddr)
+
+	return &ApplyResponse{
+		Resource: &models.Resource{
+			ID:         r.ID,
+			Type:       r.Type,
+			Attributes: r.Attributes,
+			DependsOn:  planState.DependsOn,
+			Extensions: planState.Extensions,
+		},
+		Status: nil,
+	}
+}
+
+// Read terraform show state
+func (t *TerraformRuntime) Read(ctx context.Context, request *ReadRequest) *ReadResponse {
```

Comment:
Func Read() should return the live state of this resource, not the local state file